pkg/controllers/provisioningv2/provisioningcluster/template.go

```diff
@@ -336,13 +336,10 @@ func machineDeployments(cluster *rancherv1.Cluster, capiCluster *capi.Cluster, d
 			infraRef = *machinePool.NodeConfig
 		}
 
-		if machinePool.MachineOS == "" {
-			machinePool.MachineOS = capr.DefaultMachineOS
+		machineOS := machinePool.MachineOS
+		if machineOS == "" {
+			machineOS = capr.DefaultMachineOS
 		}
-		if machinePool.MachineDeploymentLabels == nil {
-			machinePool.MachineDeploymentLabels = make(map[string]string)
-		}
-		machinePool.MachineDeploymentLabels[capr.CattleOSLabel] = machinePool.MachineOS
```

Comment:
Note to reviewer:

Although `machinePool` is a local variable, it is only a shallow copy of the pool in the spec, so it kept the reference to the same `MachineDeploymentLabels` map, as long as it already had a non-`nil` value. This is why this change had an effect on the original cluster object, which would sometimes be persisted if subsequent handlers in the controller chain updated the object.